test/tokenmanager.js

```diff
@@ -462,4 +467,106 @@ contract('Token Manager', ([root, holder, holder2, anyone]) => {
+            await tokenManager.initialize(token.address, true, 0)
+
+            hook0 = await TokenManagerHook.new(0)
+            hook1 = await TokenManagerHook.new(1)
+            hook2 = await TokenManagerHook.new(2)
+            await tokenManager.registerHook(hook0.address)
+            await tokenManager.registerHook(hook1.address)
+            await tokenManager.registerHook(hook2.address)
+            await tokenManager.revokeHook(1)
+        })
+
+        it('can register hooks', async () => {
+            const hook3 = await TokenManagerHook.new(3)
+            const { receipt } = await tokenManager.registerHook(hook3.address)
+            assert.equal(parseInt(getTopicArgument(receipt, 'RegisterHooked(uint256)', 1)), 3)
+        })
+
+        it('can revoke hooks', async () => {
+            const { receipt } = await tokenManager.revokeHook(0)
+            assert.equal(parseInt(getTopicArgument(receipt, 'RevokeHooked(uint256)', 1)), 0)
+        })
+
+        it('calls onTransfer hook on token mintings', async () => {
+            const { receipt } = await tokenManager.mint(holder, 10)
+            assert.equal(parseInt(getTopicArgument(receipt, 'TransferHooked(uint256,address,address)', 1, 0)), 0)
+            assert.equal(parseInt(getTopicArgument(receipt, 'TransferHooked(uint256,address,address)', 1, 1)), 2)
+
+            // Transfer from 0x0 to holder
+            assert.equal(getTopicArgumentAsAddr(receipt, 'TransferHooked(uint256,address,address)', 2), ZERO_ADDR)
+            assert.equal(getTopicArgumentAsAddr(receipt, 'TransferHooked(uint256,address,address)', 3), holder)
+        })
+
+        it('calls onTransfer hook on token transfers', async () => {
+            const { receipt: receipt1 } = await tokenManager.issue(10)
+            assert.equal(parseInt(getTopicArgument(receipt1, 'TransferHooked(uint256,address,address)', 1, 0)), 0)
+            assert.equal(parseInt(getTopicArgument(receipt1, 'TransferHooked(uint256,address,address)', 1, 1)), 2)
+
+            const { receipt: receipt2 } = await tokenManager.assign(holder, 5)
+            assert.equal(parseInt(getTopicArgument(receipt2, 'TransferHooked(uint256,address,address)', 1, 0)), 0)
+            assert.equal(parseInt(getTopicArgument(receipt2, 'TransferHooked(uint256,address,address)', 1, 1)), 2)
+
+            const { receipt: receipt3 } = await token.transfer(tokenManager.address, 5, { from: holder })
+            assert.equal(parseInt(getTopicArgument(receipt3, 'TransferHooked(uint256,address,address)', 1, 0)), 0)
+            assert.equal(parseInt(getTopicArgument(receipt3, 'TransferHooked(uint256,address,address)', 1, 1)), 2)
+
+            // Transfer from holder to token manager
+            assert.equal(getTopicArgumentAsAddr(receipt3, 'TransferHooked(uint256,address,address)', 2), holder)
+            assert.equal(getTopicArgumentAsAddr(receipt3, 'TransferHooked(uint256,address,address)', 3), tokenManager.address)
```

Comment:
Maybe consider checking the balance as well, although should be fine.

Reply:
I tried, but I can't index more than 3 parameters in an event, and `id`, `from`, and `to` are already indexed.

Reply:
I don't understand, you have the sender and receiver addresses, why can't you get the sender and recipient balances before and after the transaction?

Reply:
I added some checks in the last commit. I think we are ready to go.